File: src/1257-Smallest-Common-Region/1257.cpp
class Solution 
{
public:
    string findSmallestRegion(vector<vector<string>>& regions, string region1, string region2) 
    {
        unordered_map<string, string> parent;
        for (auto& region : regions) 
        {
            for (auto it = region.begin() + 1; it != region.end(); it++)
            {
                parent[*it] = region[0];
            }
        }
        string n1 = region1, n2 = region2;
        while (n1 != n2)
        {
            n1 = parent.count(n1) ? parent[n1] : region2;
            n2 = parent.count(n2) ? parent[n2] : region1;
        }
        return n1;
    }
};

File: src/1257-Smallest-Common-Region/1257.py
class Solution:
    def findSmallestRegion(self, regions: List[List[str]], region1: str, region2: str) -> str:
        parent = {r: region[0] for region in regions for r in region[1:]}
        n1, n2 = region1, region2
        while n1 != n2:
            n1 = parent[n1] if n1 in parent else region2
            n2 = parent[n2] if n2 in parent else region1
        return n1